Added tag bundles

cheriehu opened the pull request:
https://github.com/christianrauch/apriltag_ros/issues/55

This is my attempt at implementing tag bundles for this package using pnp. 

These are the parameters used for the video below. 
```
tag_bundles:
    bundle_names: ['bundle1', 'bundle2']
    bundle1:
        ids: [0, 1] # tag IDs in bundle
        0:
            size: 0.1
            transform: [-0.25, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0] # [x, y, z, qw, qx, qy, qz]
        1:
            size: 0.1
            transform: [0.25, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0] # [x, y, z, qw, qx, qy, qz]
    bundle2:
        ids: [0] # tag IDs in bundle
        0:
            size: 0.1
            transform: [-0.25, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0] # [x, y, z, qw, qx, qy, qz]

```

https://github.com/user-attachments/assets/d355c5e9-209d-4e22-b21c-be8043de7053

cheriehu committed:
multiple tag bundle

cheriehu committed:
remove apply_transform_to_point3d function from pose estimation

cheriehu committed:
change how bundle parameters are defined

cheriehu committed:
use quaternions instead of roll, pitch, yaw

cheriehu committed:
fix style issue

christianrauch commented on src/AprilTagNode.cpp:
```diff
@@ -221,12 +248,21 @@ void AprilTagNode::onCamera(const sensor_msgs::msg::Image::ConstSharedPtr& msg_i
+            bool id_in_bundle = (std::find(bundle.ids.begin(), bundle.ids.end(), det->id) != bundle.ids.end());
+            if(id_in_bundle) {
+                bundle_detections[bundle.frame_id].push_back(det);
+                // RCLCPP_INFO(get_logger(), "tag of id %i found", det->id);
```
No commented out code.

christianrauch commented on src/AprilTagNode.cpp:
```diff
@@ -251,6 +287,16 @@ void AprilTagNode::onCamera(const sensor_msgs::msg::Image::ConstSharedPtr& msg_i
+            bundle_transform_stamped.transform = pnp_bundle(bundle_detections[bundle.frame_id], intrinsics, bundle.id_to_size, bundle.id_to_tf);
+            tf_broadcaster.sendTransform(bundle_transform_stamped);
+        }
+    }
```
Same here. Use the actual type and const where possible.

christianrauch commented on src/AprilTagNode.cpp:
```diff
@@ -251,6 +287,16 @@ void AprilTagNode::onCamera(const sensor_msgs::msg::Image::ConstSharedPtr& msg_i
     }
 
+    for(auto& bundle : all_tag_bundles) {
+        geometry_msgs::msg::TransformStamped bundle_transform_stamped;
```
This line can go into the "if" block below.

christianrauch commented:
Thanks for working on this.

cheriehu committed:
use actual type and const, add EOL

move bundle transform publisher into if block

add check for whether individual tag id is already part of a bundle

use eigen for transformation

cheriehu commented on cfg/tags_36h11.yaml:
```diff
@@ -23,3 +23,20 @@
+                    transform: [-0.25, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0] # [x, y, z, qw, qx, qy, qz]
+                1:
+                    size: 0.1
+                    transform: [0.25, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0] # [x, y, z, qw, qx, qy, qz]
```
I've tested it with a cube in gazebo. These are the parameters used.

```
        # optional list of tags
        tag:
            ids: [0, 1, 2, 3]            # tag ID
            frames: ["  0", "  1", "  2", "  3"]  # optional frame name

        # optional list of tag bundles
        tag_bundles:
            bundle_names: ['bundle1']
            bundle1:
                ids: [0, 1, 2, 3] # tag IDs in bundle
                0:
                    size: 0.1
                    transform: [0.0, 0.0, 0.06, 1.0, 0.0, 0.0, 0.0] # [x, y, z, qw, qx, qy, qz]
                1:
                    size: 0.1
                    transform: [0.06, 0.0, 0.0, 0.7073883, 0.0, 0.7068252, 0.0] # [x, y, z, qw, qx, qy, qz]
                2:
                    size: 0.1
                    transform: [0.0, 0.0, -0.06, 0.0, 0.0, 1.0, 0.0] # [x, y, z, qw, qx, qy, qz]
                3:
                    size: 0.1
                    transform: [-0.06, 0.0, 0.0, -0.7073883, 0.0, 0.7068252, 0.0] # [x, y, z, qw, qx, qy, qz]

```

https://github.com/user-attachments/assets/f3afb0db-3984-4ae8-8ec7-b9f2dace4766

christianrauch reviewed (commented):
Can you squash the commits together?

christianrauch commented on src/AprilTagNode.cpp:
```diff
@@ -221,12 +249,20 @@ void AprilTagNode::onCamera(const sensor_msgs::msg::Image::ConstSharedPtr& msg_i
+            if(id_in_bundle) {
+                bundle_detections[bundle.frame_id].push_back(det);
+            }
+        }
```
Isn't this information already in `tags_in_bundles`? I suggest gathering this information once in the beginning, after parsing the bundle configuration, and then reuse it to check if a tag is a standalone tag or part of a bundle.

cheriehu replied:
Yep you're right, will change this part to use the `tags_in_bundles` variable

christianrauch commented on src/pose_estimation.cpp:
```diff
@@ -54,6 +54,50 @@ pnp(apriltag_detection_t* const detection, const std::array<double, 4>& intr, do
+            Eigen::Vector3d(+s, +s, 0),
+            Eigen::Vector3d(-s, +s, 0)};
+
+        Eigen::Affine3d transform = Eigen::Translation3d(tf[0], tf[1], tf[2]) * Eigen::Quaternion<double>(tf[3], tf[4], tf[5], tf[6]);
```
Replace `Eigen::Quaternion<double>` with `Eigen::Quaterniond`.

christianrauch reviewed (changes requested):
Can you also squash all this into a single commits? You now have many fixup changes.

christianrauch commented:
Are you still working on this? Can you check the remaining comments, address them and rebase?

josephduchesne committed:
Applied feedback from pre-existing PR

(cherry picked from commit 3bd48fbe42031686910664a04a2f9277d9235876)

josephduchesne committed:
Migrated camera matrix + pnp to its own method as requested

(cherry picked from commit 433b21fe063741d6d3330daab6b02ef574608827)

josephduchesne committed:
Fixed code style issues

(cherry picked from commit 9f83f3d49bb4c0c7a958cc1e98f3e1a633928c13)

cheriehu committed:
Merge branch 'christianrauch:master' into tag-bundle

cheriehu committed:
apply requested changes
